### PR TITLE
Make parse_primary loop back into recursive descent parser

### PR DESCRIPTION
**Pull request by mkaminski1988**
Remove redundant logic in `parse_primary()`.

This change made detecting the `Multiplying node values is not supported` check impractical because it's difficult to keep track of whether a node selector followed by a wildcard falls within an expression. 

After thinking it over, I realized that `$[?(@.key*2==100)]` is arguably a lexing error rather than a parsing error. The only allowed wildcard syntax includes the following: `.*`, `..*` and `[*]`. Based on that, `@.key*2` is invalid. I moved this check into the lexer, where it returns `Unrecognized token...`. 

I think this is a bit cleaner than explicitly forbidding multiplication (why not subtraction, etc...). Let me know what you think.

### CODE REVIEW

**Review by crocodele — APPROVED** [review]

File: src/jsonpath/lexer.c [hunk 1]
```diff
@@ -122,6 +124,15 @@ bool scan(char** p, struct jpath_token* token, char* json_path) {
             token->type = LEX_EXPR_START;
             NEXT_CHAR();
             return true;
+          case '*':
+            token->type = LEX_WILD_CARD;
+            NEXT_CHAR();
+            if (CUR_CHAR() != ']') {
+              raise_error("Wildcard filter contains an invalid chracter, expected `]`", json_path, *p);
```
Comment (crocodele) [inline src/jsonpath/lexer.c:131]: Oh, I missed a typo: `chracter` => `character`.

File: tests/comparison_union/015.phpt [hunk 2]
```diff
@@ -16,11 +16,10 @@ $data = [
 $jsonPath = new JsonPath();
 $result = $jsonPath->find($data, "$[*,1]");
 
-echo "Assertion 1\n";
 var_dump($result);
 ?>
 --EXPECTF--
-Fatal error: Uncaught RuntimeException: Missing filter end `]` in %s
+Fatal error: Uncaught RuntimeException: Wildcard filter contains an invalid chracter, expected `]` at position 3 in %s
```
Comment (crocodele) [inline tests/comparison_union/015.phpt:22]: Typo: `chracter` => `character`.

**Review by crocodele — CHANGES_REQUESTED** [review]

**Comment by crocodele** [comment]
> After thinking it over, I realized that `$[?(@.key*2==100)]` is arguably a lexing error rather than a parsing error. The only allowed wildcard syntax includes the following: `.*`, `..*` and `[*]`. Based on that, `@.key*2` is invalid. I moved this check into the lexer, where it returns `Unrecognized token...`.
> 
> I think this is a bit cleaner than explicitly forbidding multiplication (why not subtraction, etc...). Let me know what you think.

Agreed, this is better.

**Comment by mkaminski1988** [comment]
@crocodele Your latest unit tests were really helpful in uncovering some edge cases with this change. After I did the initial implementation, all the tests passed until I merged in 5d5efbc9fd885df8b725e39c15af97733dab2124.

**Review by crocodele — APPROVED** [review]
